sources:

https://pandas.pydata.org/pandas-docs/stable/10min.html

https://cloudxlab.com/blog/numpy-pandas-introduction/

https://docs.scipy.org/doc/numpy/user/quickstart.html

# Data Imputation
Data imputation refers to dealing with entries in the data that don't have a valid value. To undersand better what we mean by not having a valid value, let's take a look at the fragile families data.

In [1]:
import pandas as pd
# Let's read in the data
background = "../../ai4all_data/background.csv"
data_frame = pd.read_csv(background, low_memory=False)

# Before we look at the data, let's set the index of the data frame to be equal to the challengeID
data_frame = data_frame.set_index('challengeID')

In [2]:
# Let's now look at the first few rows and columns of the data
data_frame.head()

m1intmon  m1intyr  m1lenhr  m1lenmin  cm1twoc  cm1fint  cm1tdiff  \
challengeID                                                                     
1                  -3      NaN       -3        40      NaN        0       NaN   
2                  -3      NaN        0        40      NaN        1       NaN   
3                  -3      NaN        0        35      NaN        1       NaN   
4                  -3      NaN        0        30      NaN        1       NaN   
5                  -3      NaN        0        25      NaN        1       NaN   

             cm1natsm  m1natwt  cm1natsmx    ...      m4d9     m4e23  \
challengeID                                  ...                       
1                 NaN      NaN        NaN    ...      -3.0 -3.000000   
2                 NaN      NaN        NaN    ...      -3.0  8.473318   
3                 NaN      NaN        NaN    ...      -3.0 -3.000000   
4                 NaN      NaN        NaN    ...      -3.0 -3.000000   
5                 NaN      NaN        NaN    ...      -3.0 -3.000000   

                  f4d6       f4d7  f4d9       m5c6     m5d20  m5k10  \
challengeID                                                           
1            -3.000000  -3.000000  -3.0  -3.000000 -3.000000     -3   
2            -3.000000  -3.000000  -3.0  -3.000000  9.845074     -3   
3             9.097495  10.071504  -3.0  -3.000000 -3.000000     -3   
4             9.512706  10.286578  -3.0  10.677285 -3.000000     -3   
5            11.076016   9.615958  -3.0   9.731979 -3.000000     -3   

                  f5c6       k5f1  
challengeID                        
1            -3.000000  -3.000000  
2            -3.000000   9.723551  
3            -3.000000  -3.000000  
4             8.522331  10.608137  
5            10.115313   9.646466  

[5 rows x 12942 columns]

## What are the NaN values?
NaN (Not a Number) entries appear in real-world datasets wery often, usually signifying missing data. NaNs are also produced when dividing by zero, or casting a non-numerical value to a number.

## What should we do about missing values?
Some researchers simply discard data samples where NaN values are present. This is problematic, because in relatively small datasets, this means getting rid of a large portion of the data.

The alternative solution is to *impute* - or fill in - missing data points. However, correct imputation requires advanced statistical knowledge. Sometimes, the average of a given column is used to replace NaN values. Other times, values are copied from other rows which have similar entries in the non-missing columns (the K Nearest Neighbors algorithm).

During this project, we will use three ways of dealing with missing data:
* removing NaN columns (simplest)
* filling in average values of the column (potentially making a simplifying assumption)
* using the *K Nearest Neighbors (KNN)* algorithm described above

### 1. Removing NaN values
We can either remove all the columns (*features*) that contain NaNs, or all the rows (*data points*)

In [18]:
# The isnull() function checks item == NaN for each item in a data frame and returns a boolean valued data frame
is_it_null = data_frame.isnull()
is_it_null.head()

m1intmon  m1intyr  m1lenhr  m1lenmin  cm1twoc  cm1fint  cm1tdiff  \
challengeID                                                                     
1               False     True    False     False     True    False      True   
2               False     True    False     False     True    False      True   
3               False     True    False     False     True    False      True   
4               False     True    False     False     True    False      True   
5               False     True    False     False     True    False      True   

             cm1natsm  m1natwt  cm1natsmx  ...     m4d9  m4e23   f4d6   f4d7  \
challengeID                                ...                                 
1                True     True       True  ...    False  False  False  False   
2                True     True       True  ...    False  False  False  False   
3                True     True       True  ...    False  False  False  False   
4                True     True       True  ...    False  False  False  False   
5                True     True       True  ...    False  False  False  False   

              f4d9   m5c6  m5d20  m5k10   f5c6   k5f1  
challengeID                                            
1            False  False  False  False  False  False  
2            False  False  False  False  False  False  
3            False  False  False  False  False  False  
4            False  False  False  False  False  False  
5            False  False  False  False  False  False  

[5 rows x 12942 columns]

In [19]:
removed_rows = data_frame.dropna(axis=0)
removed_rows

Empty DataFrame
Columns: [m1intmon, m1intyr, m1lenhr, m1lenmin, cm1twoc, cm1fint, cm1tdiff, cm1natsm, m1natwt, cm1natsmx, m1natwtx, cm1citsm, m1citywt, innatsm, incitysm, m1a3, m1a4, m1a5, m1a6, m1a7, m1a8, m1a9, m1a10, m1a11a, m1a11b, m1a11c, m1a11d, m1a12, m1a12a, m1a13, m1a13a, m1a15, m1a16, cm1age, cm1bsex, cm1lbw, cm1numb, m1b0, m1b1a, m1b1b, m1b2, m1b3, m1b4a, m1b4b, m1b4c, m1b4d, m1b4e, m1b4f, m1b4g, m1b5a, m1b5b, m1b5c, m1b5d, m1b6a, m1b6b, m1b6c, m1b6d, m1b6e, m1b6f, m1b7a, m1b7b, m1b7c, m1b7d, m1b7e, m1b8, m1b9, m1b9a1, m1b9a2, m1b9a3, m1b9a4, m1b9a5, m1b9a6, m1b9a7, m1b9a8, m1b9a9, m1b9a10, m1b9a11, m1b9b1, m1b9b2, m1b9b3, m1b9b4, m1b9b5, m1b9b6, m1b9b7, m1b9b8, m1b9b9, m1b9b10, m1b9b11, m1b10, m1b10a1, m1b10a2, m1b10a3, m1b10a4, m1b10a5, m1b10a6, m1b10a7, m1b10a8, m1b10a9, m1b10a10, m1b10a11, ...]
Index: []

[0 rows x 12942 columns]

In [20]:
removed_columns = data_frame.dropna(axis=1)
removed_columns.head()

m1intmon  m1lenhr  m1lenmin  cm1fint  cm1citsm  innatsm  \
challengeID                                                            
1                  -3       -3        40        0         1        1   
2                  -3        0        40        1         1        1   
3                  -3        0        35        1         1        1   
4                  -3        0        30        1         1        1   
5                  -3        0        25        1         1        1   

             incitysm  m1a3  m1a4  m1a5    ...      m4d9     m4e23       f4d6  \
challengeID                                ...                                  
1                   1     1     1    -6    ...      -3.0 -3.000000  -3.000000   
2                   1     2     2     1    ...      -3.0  8.473318  -3.000000   
3                   1     1     1    -6    ...      -3.0 -3.000000   9.097495   
4                   1     1     2     1    ...      -3.0 -3.000000   9.512706   
5                   1     1     2     1    ...      -3.0 -3.000000  11.076016   

                  f4d7  f4d9       m5c6     m5d20  m5k10       f5c6       k5f1  
challengeID                                                                     
1            -3.000000  -3.0  -3.000000 -3.000000     -3  -3.000000  -3.000000  
2            -3.000000  -3.0  -3.000000  9.845074     -3  -3.000000   9.723551  
3            10.071504  -3.0  -3.000000 -3.000000     -3  -3.000000  -3.000000  
4            10.286578  -3.0  10.677285 -3.000000     -3   8.522331  10.608137  
5             9.615958  -3.0   9.731979 -3.000000     -3  10.115313   9.646466  

[5 rows x 9852 columns]

In the first case, our DataFrame became empty. There are no rows where all the columns are nonmissing.

In the second case, the size of the DataFrame changed, but not dramatically. Notice that the number of features dropped from 12942 to 9852.

However, since we will only be working with a few of the 12942 features at any time, we do not need to worry about removing the NaNs consistently in the entire table right away. 

In [21]:
# Select all rows, and columns 0-2 (:3)
first_three = data_frame[["m4d9","m4e23","f4d6"]]
print(first_three.head())
print(first_three.shape)
# Now, remove the individuals who do not have all three features
X = first_three[(~first_three.isnull()).all(1)]
print(X.head())
print(X.shape)

             m4d9     m4e23       f4d6
challengeID                           
1            -3.0 -3.000000  -3.000000
2            -3.0  8.473318  -3.000000
3            -3.0 -3.000000   9.097495
4            -3.0 -3.000000   9.512706
5            -3.0 -3.000000  11.076016
(4242, 3)
             m4d9     m4e23       f4d6
challengeID                           
1            -3.0 -3.000000  -3.000000
2            -3.0  8.473318  -3.000000
3            -3.0 -3.000000   9.097495
4            -3.0 -3.000000   9.512706
5            -3.0 -3.000000  11.076016
(4242, 3)


### 2. Filling the mean values in each column
You can use a pandas function fillna:

https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.fillna.html

The function may take a while to run, since it has to calculate the mean of almost 13,000 columns and then fill in all the NaN values.

In [68]:
# It looks like some columns are completely empty - filled fully with NaNs...
print(data_frame['m1intyr'].isnull().sum())
print(data_frame['cm1twoc'].isnull().sum())
print(data_frame['m1natwt'].isnull().sum())

4242
4242
4242


In [62]:
fill_mean = data_frame.fillna(data_frame.mean())

In [70]:
# Since some columns are NaNs, this didn't exactly work...
fill_mean.head()

m1intmon  m1intyr  m1lenhr  m1lenmin  cm1twoc  cm1fint  cm1tdiff  \
challengeID                                                                     
1                  -3      NaN       -3        40      NaN        0       NaN   
2                  -3      NaN        0        40      NaN        1       NaN   
3                  -3      NaN        0        35      NaN        1       NaN   
4                  -3      NaN        0        30      NaN        1       NaN   
5                  -3      NaN        0        25      NaN        1       NaN   

             cm1natsm  m1natwt  cm1natsmx    ...      m4d9     m4e23  \
challengeID                                  ...                       
1                 NaN      NaN        NaN    ...      -3.0 -3.000000   
2                 NaN      NaN        NaN    ...      -3.0  8.473318   
3                 NaN      NaN        NaN    ...      -3.0 -3.000000   
4                 NaN      NaN        NaN    ...      -3.0 -3.000000   
5                 NaN      NaN        NaN    ...      -3.0 -3.000000   

                  f4d6       f4d7  f4d9       m5c6     m5d20  m5k10  \
challengeID                                                           
1            -3.000000  -3.000000  -3.0  -3.000000 -3.000000     -3   
2            -3.000000  -3.000000  -3.0  -3.000000  9.845074     -3   
3             9.097495  10.071504  -3.0  -3.000000 -3.000000     -3   
4             9.512706  10.286578  -3.0  10.677285 -3.000000     -3   
5            11.076016   9.615958  -3.0   9.731979 -3.000000     -3   

                  f5c6       k5f1  
challengeID                        
1            -3.000000  -3.000000  
2            -3.000000   9.723551  
3            -3.000000  -3.000000  
4             8.522331  10.608137  
5            10.115313   9.646466  

[5 rows x 12942 columns]

### 3. Using the K Nearest Neighbors (KNN) algorithm 

In [22]:
#married[married==True].index.values
#array[~np.isnan(bg.cm1age)]